### PR TITLE
FeatureInfoAgent v2.0.0

### PR DESCRIPTION
**Pull request by mdhillman**
This PR brings in major changes to the FeatureInfoAgent. These changes include:

- Separation into multiple classes
- Addition of JUnit Tests
- Updated documentation
- Seperation of queries for metadata and timeseries
- Automatic discovery and federation of Blazegraph endpoints

For more details please see the associated README file.

### CODE REVIEW

**Comment by mdhillman** [comment]
This PR is a draft and will not be published until the following conditions are met:

- The stack client library is merged to make the getServiceUrl() is available on a non-SNAPSHOT version.
- The agent has been tested within a real stack, with real data.
  - This has not been possible during development as the overhead of creating a mock stack with mock KG and RDBs is too high.
- A GitHub action have been added to run units tests upon PR submission.
- A GitHub action has been added to build and push the docker image after a PR is merged.

File: Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/Utils.java
Comment (kokfoongcmcl) [inline Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/Utils.java:36]: This is specific to the river sensor stations due to the way I initialised them, one station may contain multiple tables, typically one station will contain a table with multiple columns if it measures more than 1 thing. It's a combination of laziness and how the API returns data

File: Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/FeatureInfoAgent.java
Comment (kokfoongcmcl) [inline Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/FeatureInfoAgent.java:52]: Seb will disagree on this, but there's no point in using JPSAgent if you're not using processRequestParameters

File: Agents/FeatureInfoAgent/code/WEB-INF/web.xml
Comment (kokfoongcmcl) [inline Agents/FeatureInfoAgent/code/WEB-INF/web.xml:1]: I believe this file is redundant when you use the @WebServlet annotation in your FeatureInfoAgent class

File: Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/FeatureInfoAgent.java
Comment (kokfoongcmcl) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/FeatureInfoAgent.java:52]: Ok I see now that you are using readJsonParameter, then this is not applicable, documentation somewhere needs to say this takes in a "query" parameter with a JSON string in it

File: Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/Utils.java
Comment (mdhillman) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/Utils.java:36]: If this code is run on non-river sensor stations what will happen? 

This agent is only going to be able to handle timeseries if timeseries are storred generically for all station/location types. I don't really want to be adding an ever growing list of "if(thisTypeOfStation){}" clauses.

File: Agents/FeatureInfoAgent/code/WEB-INF/web.xml
Comment (mdhillman) [reply Agents/FeatureInfoAgent/code/WEB-INF/web.xml:1]: You may be right, I'll make a note to try removing it in future (it's not critical to do so now).

File: Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/Utils.java
Comment (kokfoongcmcl) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/Utils.java:36]: Based on the way you create the time series objects it won't make any difference, but it's a bit inefficient

File: Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java
Comment (kokfoongcmcl) [inline Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java:347]: In the latest base lib, if you don't pass a connection object the these methods, it will open and close a connection for each query you make, if you want to make it faster, pass a connection object and close it at the end using try-with-resource

File: Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/Utils.java
Comment (mdhillman) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/Utils.java:36]: I'd suggest we leave it as-is then. For this PR I'm more interested in getting the functionality correct than it being more efficient.

File: Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java
Comment (mdhillman) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java:347]: This is something I can add in. Is there a method in the base lib to create the initial connection, or do I just use generic JDBC code?

Comment (kokfoongcmcl) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java:347]: There's a RemoteRDBStoreClient.getConnection() method, which is basically calling the jdbc code

**Comment by mdhillman** [comment]
> This PR is a draft and will not be published until the following conditions are met:
> 
> * The stack client library is merged to make the getServiceUrl() is available on a non-SNAPSHOT version.
> * The agent has been tested within a real stack, with real data.
>   
>   * This has not been possible during development as the overhead of creating a mock stack with mock KG and RDBs is too high.
> * A GitHub action has been added to run units tests upon PR submission.
> * A GitHub action has been added to build and push the docker image after a PR is merged.

The two mentioned GitHub actions have now been added.

File: Agents/FeatureInfoAgent/Dockerfile
Comment (sm453) [inline Agents/FeatureInfoAgent/Dockerfile:11]: since encrypted passwords can contain the LATTER, but not the FORMER

File: Agents/FeatureInfoAgent/code/pom.xml
Comment (sm453) [inline Agents/FeatureInfoAgent/code/pom.xml:60]: The comment says "with Java 11"...

File: Agents/FeatureInfoAgent/README.md
Comment (markushofmeister) [reply Agents/FeatureInfoAgent/README.md:92]: But I guess building would not _always_ be required and people should rather just pull the image, right?

**Review by markushofmeister — CHANGES_REQUESTED** [review]
I am unable to build (and push) the Docker image using `bash ./stack.sh build`. I am getting a `denied: permission_denied: write_package` error despite being able to push other Docker images to ghcr.io (from the same machine). The latest version of the published image is 4 days old. Is this still the latest and worth testing the functionality with it?

File: Agents/FeatureInfoAgent/README.md
Comment (gpeb2) [reply Agents/FeatureInfoAgent/README.md:22]: I think Michael means the official HTTP response code 204 "NO_CONTENT" ([List_of_HTTP_status_codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)). I'm not suggesting there isn't a better code though.

File: Agents/FeatureInfoAgent/README.md
Comment (gpeb2) [reply Agents/FeatureInfoAgent/README.md:92]: Yes, building should _only_ be done if changes have been made to that code.

Comment (sm453) [reply Agents/FeatureInfoAgent/README.md:92]: Correct.

File: Agents/FeatureInfoAgent/README.md
Comment (mdhillman) [reply Agents/FeatureInfoAgent/README.md:22]: Yes, I'd like to stick to approved HTTP codes if possible. I think NO_CONTENT conveys that the connection to the KG _was_ made okay, but it just didn't contain details on the class; but I'm open to suggestions.

File: Agents/FeatureInfoAgent/README.md
Comment (mdhillman) [reply Agents/FeatureInfoAgent/README.md:26]: The last think the FIA does is takes that `property`, `value`, `unit` structure that is expected from queries and parses it to something a little more human-readable. For instance...

```json
[{
  "Property": "Length",
  "Value": "100",
  "Unit": "m" 
}]
```

...would get parsed into the following before being passed back to the visualisation.

```json
[{
  "Length": "100 [m]"
]}
```

File: Agents/FeatureInfoAgent/README.md
Comment (mdhillman) [reply Agents/FeatureInfoAgent/README.md:43]: First three points are across everything the FIA does right now, whether the queries are federated or not.

The last point is more of a user facing requirement; in theory the FIA can handle as large a query as the base lib can, but we don't want the user to be waiting within the visualisation for minutes.

File: Agents/FeatureInfoAgent/README.md
Comment (mdhillman) [reply Agents/FeatureInfoAgent/README.md:33]: That does sound like a reasonable idea. Would a query to just see if an IRI exists within a 8-10 Mil triplestore be lengthy?

File: Agents/FeatureInfoAgent/README.md
Comment (mdhillman) [reply Agents/FeatureInfoAgent/README.md:79]: Yes, I believe so.

File: Agents/FeatureInfoAgent/README.md
Comment (mdhillman) [reply Agents/FeatureInfoAgent/README.md:92]: I have updated the README, but if anyone has any details on how to update the stack routing then it may be worth adding/linking to that.

**Comment by mdhillman** [comment]
Connection object has been added for the timeseries RDB queries, will need deployment within a stack to test if this improves the efficiency.

**Comment by mdhillman** [comment]
> I am unable to build (and push) the Docker image using `bash ./stack.sh build`. I am getting a `denied: permission_denied: write_package` error despite being able to push other Docker images to ghcr.io (from the same machine). The latest version of the published image is 4 days old. Is this still the latest and worth testing the functionality with it?
> 
> Is this due to the fact that publishing the image is now done by github actions? Just confused as the latest image is much older than your tag `ready for review`.

How the image is push shouldn't affect who else can push it (as far as I'm aware). There is a kink with uploading images to GitHub though; the first time it's done they appear to default to private. I believe I've changed the settings to make the FIA image public, but I will double check.

**Comment by mdhillman** [comment]
> > I am unable to build (and push) the Docker image using `bash ./stack.sh build`. I am getting a `denied: permission_denied: write_package` error despite being able to push other Docker images to ghcr.io (from the same machine). The latest version of the published image is 4 days old. Is this still the latest and worth testing the functionality with it?
> > Is this due to the fact that publishing the image is now done by github actions? Just confused as the latest image is much older than your tag `ready for review`.
> 
> How the image is push shouldn't affect who else can push it (as far as I'm aware). There is a kink with uploading images to GitHub though; the first time it's done they appear to default to private. I believe I've changed the settings to make the FIA image public, but I will double check.

I had made it public, but only for download. It looks like there's a second stage (that I missed) that allows permissions inheritance from the source repository. I've enabled that now so everyone should have the same access that they have to the TWA code.

For now, you may need to build the image locally. The current actions will only push it _after_ the PR is closed. As this doesn't help you test it, what I should really do it have the PR push a SNAPSHOT version then a non-SNAPSHOT once it's approved. I'll need to discuss with @George on how to achieve this.

File: Agents/FeatureInfoAgent/README.md
Comment (markushofmeister) [reply Agents/FeatureInfoAgent/README.md:22]: Sounds sensible. It was not clear to me that your refer to official HTTP response codes here

File: Agents/FeatureInfoAgent/README.md
Comment (markushofmeister) [reply Agents/FeatureInfoAgent/README.md:26]: Makes perfect sense; "approved" could be elaborated here for clarify, e.g. Parse resulting metadata from SPARQL query into visualisation ready format

File: Agents/FeatureInfoAgent/README.md
Comment (markushofmeister) [reply Agents/FeatureInfoAgent/README.md:33]: It should not. Queries for specific IRIs should be very quick generally; however, depending on the metadata query and in case an IRI exists in multiple namespaces, this could lead to lengthy queries in case the query is federated to the unnecessary endpoint first.

File: Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java
Comment (kokfoongcmcl) [inline Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java:112]: It looks like you are storing the connection object, it's not recommended, you should do something like
`try (Connection conn = rdbClient.getConnection()) {
    timeSerieClient.getTimeSeries(....., conn)
}`

File: Agents/FeatureInfoAgent/README.md
Comment (mdhillman) [reply Agents/FeatureInfoAgent/README.md:33]: Unless we find the current implementation too slow, then I suggest we leave this to be looked into as a new feature within a subsequent version of the agent.

**Comment by sm453** [comment]
For time series queries, would it be possible to make each of the `queries` have its own `hours` parameter (in the config json)? A use case may contain many time series of substantially different scale (e.g. hours vs. years).

**Comment by mdhillman** [comment]
> For time series queries, would it be possible to make each of the `queries` have its own `hours` parameter (in the config json)? A use case may contain many time series of substantially different scale (e.g. hours vs. years).

Yes this should be possible (although not trivial), I will take a look.

**Comment by mdhillman** [comment]
> > For time series queries, would it be possible to make each of the `queries` have its own `hours` parameter (in the config json)? A use case may contain many time series of substantially different scale (e.g. hours vs. years).
> 
> Yes this should be possible (although not trivial), I will take a look.

This change has now been implemented, a `timeLimit` parameter can be set for each class to set an limit for timeseries data (in hours, defaults to 24 if not present).

File: Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java
Comment (kokfoongcmcl) [inline Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java:373]: Are we sure that this is what is required? Should it be something like taking the latest entry in the time series table and subtracting that instead?

Comment (sm453) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java:373]: I'm inclined to agree with @kokfoongcmcl here. It is likely more useful to see the so many _most recent_ hours, rather than so many hours back from _now_ (which is likely to be nothing at all very quickly!).

Comment (mdhillman) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java]: The current design we've used in all previous visualisations is to show data from the last 24 hours.

We can change this to range from N hours since the latest data point, but this is something I would require to be done within the TimeseriesClient (I suggest as a new method). Unless some derivation calculation is being made, I don't think agents should handle filtering the timeseries at all, this should be done centrally.

Comment (kokfoongcmcl) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java]: The easiest thing to do now is to use getLatestData, get the value of the time from the TimeSeries object, then calculate the range and use getTimeSeriesWithinBounds.

It might be tricky to implement this in the TimeSeriesClient itself, because I allowed the Time class to be anything (e.g. Integer, Double, etc.), if it's Double, and you provide a Duration object, I don't know how to deal with it

Comment (mdhillman) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java]: Would this trigger 2 separate connections to the RDB?

I strongly think the TimeseriesClient should be responsible for any filtering, as I don't want to tie any agents directly to the format in which timeseries are returned. Additionally, this sounds like something we'll want to use in different agents, so should be done centrally.

Comment (kokfoongcmcl) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java]: You can use the same connection object for both calls.

It's not straightforward to support that, it will need to have one method for each class, I can't just do "-" unless they're subclass of Number

Comment (mdhillman) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java]: I suggest this request goes on the back burner then until such time that it can be added centrally.

Comment (kokfoongcmcl) [reply Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java]: As Seb said, we don't think this is a good idea anyway because you'll get empty results quite quickly, depending on the data of course

**Comment by sm453** [comment]
> For time series queries, would it be possible to make each of the `queries` have its own `database_name` parameter (in the config json)? In a complex use case it may not be practical to have all time series in the same database, hence a single, global setting is a bit restrictive.

Or even better, read the hasRDB property from the KG, but whatever is quicker to implement wins for now...

**Comment by kokfoongcmcl** [comment]
> > For time series queries, would it be possible to make each of the `queries` have its own `database_name` parameter (in the config json)? In a complex use case it may not be practical to have all time series in the same database, hence a single, global setting is a bit restrictive.
> 
> Or even better, read the hasRDB property from the KG, but whatever is quicker to implement wins for now...

just have to be careful, the URL is probably the one accessed by the agent instantiating the time series, not the external one

**Comment by markushofmeister** [comment]
> I'm not sure if _all_ data has the `hasRDB` relationship yet, pretty sure the data on our development servers (which I use for testing) doesn't. Until that happens, I'd be wary of relying on it.

I might be mistaken, but I do not see why earlier timeseries should not have the `hasRDB` relationship. I just checked again, and `hasRDB` has been part of the initial commit of the TimeSeries Ontology

**Comment by sm453** [comment]
I just checked the triples generated by a few current instantiation agents, they all put the stack-internal JDBC URLs, which is exactly what we need. However, if it is easier/quicker/less problematic for now to allow the config to specify a DB name per class, then that is the way to go. Better solutions can be left for later.

**Review by markushofmeister — CHANGES_REQUESTED** [review]
After adding `<script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.29.4/moment.min.js" ></script>` to the `index.html` I am able to visualise retrieved meta and timeseries data. However, retrieving time series data still causes issues in cases where units have encoding problems.
As "wrong" instantiation of non-ascii characters within Blazegraph seems to be a known issue (see issue #513), we should try to have a robust solution for that.

To reproduce the issue, please use the MetOffice stations `Marham` (airfield south of King's Lynn, `<https://www.theworldavatar.com/kg/ontoems/ReportingStation_897428dc-6340-4c56-b632-50bd7f128f61>`) at the endpoint `http://165.232.172.16:3838/blazegraph/ui/#query` (namespace `stations`). I am using the timeseries.sparql file as below.
Without commenting out both `  ?Measurement ^om:hasValue/rdf:type <https://www.theworldavatar.com/kg/ontoems/RelativeHumidity> . `, I can retrieve the one time series successfully. However, when trying to retrieve all time series by commenting those two lines out, the time series panel remains completely empty, although Blazegraph returns "correct" results:
image

For other stations, time series with wrongly encoded units are shown as below:
image


```
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
prefix om:  <http://www.ontology-of-units-of-measure.org/resource/om-2/>
prefix ems: <https://www.theworldavatar.com/kg/ontoems/>

SELECT ?Measurement ?Name ?Unit
WHERE {
  {
  # Observations/Measurements
  [IRI] ems:reports/om:hasValue ?Measurement . 
  ?Measurement ^om:hasValue/rdf:type <https://www.theworldavatar.com/kg/ontoems/RelativeHumidity> . 
  ?Measurement ^om:hasValue/rdf:type ?type ; 
               om:hasUnit/om:symbol ?Unit .                
  BIND (concat(replace(str(?type), str(ems:), ''), ' observation') as ?Name)               
  # Forecasts
  } UNION {
  [IRI] ems:reports/ems:hasForecastedValue ?Measurement . 
  ?Measurement ^om:hasValue/rdf:type <https://www.theworldavatar.com/kg/ontoems/RelativeHumidity> . 
  ?Measurement ^ems:hasForecastedValue/rdf:type ?type ; 
               om:hasUnit/om:symbol ?Unit .                
  BIND (concat(replace(str(?type), str(ems:), ''), ' forecast') as ?Name)    
  }
}
```

**Comment by mdhillman** [comment]
> https://www.theworldavatar.com/kg/ontoems/ReportingStation_897428dc-6340-4c56-b632-50bd7f128f61

I've tried reproducing this with the given stack URL and IRI, unfortunatly I cannot get the FIA to return any timeseries data (the error "Could not produce JSON objects from TimeSeries instances!" is given).

Mocking the repsonse from the agent with a JSON string previously sent to me by @markushofmeister does not reproduce the issue either. All timeseries (even those with degrees in the name) appear to work fine.

**Comment by markushofmeister** [comment]
> I've tried reproducing this with the given stack URL and IRI, unfortunatly I cannot get the FIA to return any timeseries data (the error "Could not produce JSON objects from TimeSeries instances!" is given).

I believe this (unfortunately) needs further investigation before closing this PR. I just had another try and can "confirm" that

1) There are issues with retrieving particular time series (currently, this seems to affect primarily time series with non-ascii units which might be "wrongly" encoded), e.g. for `Tottenhill` some time series get displayed correctly, while others are not at all ...

Properly shown timeseries            | "Undefined" timeseries
:-------------------------:|:-------------------------:
image |  image

... despite the time series data seems to be available for all time series in PostGIS



2) Some stations are not "detectable" by FIA at all, e.g. for `Marham` the FIA returns the fallback values from PostGIS despite both `sparql` queries seem to work when executed in Blazegraph directly
image

Metadata query            |  Timeseries query
:-------------------------:|:-------------------------:
image  |  image

**Comment by sm453** [comment]
Another request: It would be great if the metadata query in the config json was optional, too (with the front-end displaying just the PostGIS metadata if no such query is provided). This would allow attaching time series to datasets that have metadata from PostGIS, without having to write an OBDA mapping and a metadata query that pulls out the metadata that would be displayed otherwise anyway.

**Comment by mdhillman** [comment]
I have updated the FeatureInfoAgent, and the DTVF (deployed as v3.3.4) on this branch to address the below issues. The earlier these can be tested and confirmed as fixed (or re-addressed), the sooner this branch can be merged (I would like this to be done ASAP so that upcoming CReDo work doesn't have to be based on a feature branch).

- Metadata queries should now be optional, simply omit the "metaFile" entry for a class in the configuration file.
- Timeseries should no longer combine time values, this was due to older code that was created when some assumptions still held true.
- Issue #513 should now be resolved, fixing issues with the units in timeseries causing issues.

File: Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java
Comment (sm453) [inline Agents/FeatureInfoAgent/code/src/main/java/com/cmclinnovations/featureinfo/kg/TimeHandler.java:196]: Please fix the message string. A funny drag-and-drop accident, I guess...

File: web/digital-twin-vis-framework/example-cesium-vis/webspace/index.html
Comment (sm453) [inline web/digital-twin-vis-framework/example-cesium-vis/webspace/index.html:141]: Please remove.

File: web/digital-twin-vis-framework/example-cesium-vis/webspace/data.json
Comment (sm453) [inline web/digital-twin-vis-framework/example-cesium-vis/webspace/data.json:15]: You may want to remove this.

File: web/digital-twin-vis-framework/example-cesium-vis/webspace/data.json
Comment (sm453) [inline web/digital-twin-vis-framework/example-cesium-vis/webspace/data.json:35]: You may want to remove this.

File: web/digital-twin-vis-framework/example-cesium-vis/webspace/settings.json
Comment (sm453) [inline web/digital-twin-vis-framework/example-cesium-vis/webspace/settings.json:10]: You may want to remove this.

**Review by markushofmeister — APPROVED** [review]
I tested the FeatureInfoAgent 2.0 with DTVF 3.3.4 (including UTF-8 charset) for the latest MetOffice agent data in the Mapbox visualisation and can confirm that
* Meta-data gets correctly retrieved and displayed
* Time series data gets correctly retrieved and displayed, even for time series with non-ascii characters in units

**Review by sm453 — APPROVED** [review]
Tested the following (with the Cesium version of the DTVF only):
- different databases for different queries
- metadata query only, time series query only, both queries, neither
- sizing of the pop-up window for named features
All working.

**Pull request merged**
